### PR TITLE
Floating Subnet Config: Wildcards and Defaults

### PR DESCRIPTION
**Pull request by dkistner**
**How to categorize this PR?**

/kind enhancement
/priority normal
/platform openstack

**What this PR does / why we need it**:
Allow to configure a default subnet for floating pool configurations in the CloudProfile.
This will be used to assign the router interface of a Shoot cluster (if not an existing router is used) to the matching subnet.
Furthermore enable the selection of floating pool subnets by wildcards.

In addition it is now forbidden to change the floating pool or the floating pool subnet configuration as this is not supported by Neutron.

This change require a Terraformer with openstack provider `v1.37` or higher. 
The floating subnet id field is removed from the Infrastructure state as it was unused.

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener-extension-provider-openstack/issues/219

**Special notes for your reviewer**:

**Release note**:

```feature operator
Allow to configure a default subnet for floating pools configuration in the CloudProfile and enable the selection of floating pool subnet via a name pattern.
```

```feature user
The selection of the subnet in floating pool network via `InfrastructureConfig.floatingPoolSubnetName` can now be also done via a name pattern matching.
```

**Will keep it as draft as there is a final test ongoing.**

/invite @kon-angelo @MartinWeindel 

cc @RaphaelVogel  @kayrus

### CODE REVIEW

File: pkg/apis/openstack/helper/helper.go [hunk 1]
```diff
@@ -108,3 +109,70 @@ func FindKeyStoneURL(keyStoneURLs []api.KeyStoneURL, keystoneURL, region string)
...
 	return "", fmt.Errorf("cannot find keystone URL for region %q", region)
 }
+
+// FindFloatingPool receives a list of floating pools and tries to find the best
+// match for a given `floatingPoolNamePattern` considering contraints like
+// `region` and `domain`. If no matching floating pool was found then an error will be returned.
+func FindFloatingPool(floatingPools []api.FloatingPool, floatingPoolNamePattern, region string, domain *string) (*api.FloatingPool, error) {
+	var (
+		floatingPoolCandidate        *api.FloatingPool
+		maxCanidateScore             int
```
Comment (MartinWeindel) [inline pkg/apis/openstack/helper/helper.go:119]:
```suggestion
		maxCandidateScore             int
```

File: pkg/apis/openstack/helper/helper.go [hunk 2]
```diff
@@ -108,3 +109,70 @@ func FindKeyStoneURL(keyStoneURLs []api.KeyStoneURL, keystoneURL, region string)
 
 	return "", fmt.Errorf("cannot find keystone URL for region %q", region)
 }
+
+// FindFloatingPool receives a list of floating pools and tries to find the best
+// match for a given `floatingPoolNamePattern` considering contraints like
```
Comment (MartinWeindel) [inline pkg/apis/openstack/helper/helper.go:114]:
```suggestion
// match for a given `floatingPoolNamePattern` considering constraints like
```

File: pkg/apis/openstack/types_infrastructure.go [hunk 3]
```diff
@@ -25,7 +25,8 @@ type InfrastructureConfig struct {
 	metav1.TypeMeta
 	// FloatingPoolName contains the FloatingPoolName name in which LoadBalancer FIPs should be created.
 	FloatingPoolName string
-	// FloatingPoolSubnetName contains the name of a subnet in the Floating IP Pool where the router should be attached to.
+	// FloatingPoolSubnetName contains the fixed name of subnet or matching name pattern for subnet
+	// in the FLoating IP Pool where the router should be attached to.
```
Comment (MartinWeindel) [inline pkg/apis/openstack/types_infrastructure.go:29]:
```suggestion
	// in the Floating IP Pool where the router should be attached to.
```

File: pkg/apis/openstack/v1alpha1/types_infrastructure.go [hunk 4]
```diff
@@ -26,7 +26,8 @@ type InfrastructureConfig struct {
 	metav1.TypeMeta `json:",inline"`
 	// FloatingPoolName contains the FloatingPoolName name in which LoadBalancer FIPs should be created.
 	FloatingPoolName string `json:"floatingPoolName"`
-	// FloatingPoolSubnetName contains the name of a subnet in the Floating IP Pool where the router should be attached to.
+	// FloatingPoolSubnetName contains the fixed name of subnet or matching name pattern for subnet
+	// in the FLoating IP Pool where the router should be attached to.
```
Comment (MartinWeindel) [inline pkg/apis/openstack/v1alpha1/types_infrastructure.go:30]:
```suggestion
	// in the Floating IP Pool where the router should be attached to.
```

**Review by MartinWeindel — CHANGES_REQUESTED** [review]
lgtm, I only found some minor typos.

**Comment by dkistner** [comment]
Thanks @MartinWeindel done :)

**Review by MartinWeindel — APPROVED** [review]
/lgtm

File: pkg/apis/openstack/helper/helper.go [hunk 5]
```diff
@@ -108,3 +109,70 @@ func FindKeyStoneURL(keyStoneURLs []api.KeyStoneURL, keystoneURL, region string)
...
+		var fip = f
+
+		// Check non constraining floating pools with second priority
+		// which means only when no other floating pool is matching.
+		if fip.NonConstraining != nil && *fip.NonConstraining {
+			nonConstrainingFloatingPools = append(nonConstrainingFloatingPools, fip)
+			continue
+		}
+
+		if candiate, score := checkFloatingPoolCandidate(&fip, floatingPoolNamePattern, region, domain); candiate != nil && score > maxCandidateScore {
```
Comment (kon-angelo) [inline pkg/apis/openstack/helper/helper.go:133]: `candidate` typo

Comment (dkistner) [reply pkg/apis/openstack/helper/helper.go]: done. Thanks

**Pull request merged**